tls/s2n_cipher_suites.c

```diff
@@ -1155,17 +1178,187 @@ int s2n_set_cipher_as_client(struct s2n_connection *conn, uint8_t wire[S2N_TLS_C
+        /* If the kex is not supported continue to the next candidate */
+        bool kex_supported = false;
+        if (s2n_result_is_error(s2n_kex_supported(potential_match, conn, &kex_supported))) {
+            return false;
+        }
+
+        if (!kex_supported) {
+            return false;
+        }
+
+        /* If the kex is not configured correctly continue to the next candidate */
+        if (s2n_result_is_error(s2n_configure_kex(potential_match, conn))) {
+            return false;
+        }
+    }
+
+    /**
+    *= https://tools.ietf.org/rfc/rfc8446#section-4.2.11
+    *# The server MUST ensure that it selects a compatible PSK
+    *# (if any) and cipher suite.
+    **/
+    if (conn->psk_params.chosen_psk != NULL) {
+        if (potential_match->prf_alg != conn->psk_params.chosen_psk->hmac_alg) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+/**
+ * Used by the server to select a mutually-supported cipher suite.
+ * This function iterates over the server's cipher suite preference list, choosing the first 
+ * suite that is also supported by the client. If an in-group deliminator is encountered, for 
+ * each in-group cipher suite, we iterate through the client's preferences and save the 
+ * index of the client's cipher suite, throwing out the previous index if it is 
+ * less-preferred by the client than the current index. Once the end-group deliminator is 
+ * encountered, the server either accepts the idenfitied cipher suite or if no cipher suite 
+ * has been found, continues iterating through its list. This function checks that the potential cipher
+ * suite match is valid and can be used for the connection.
+ * 
+ */
+static S2N_RESULT s2n_get_mutually_supported_cipher(struct s2n_connection* conn,
+                                                    const uint8_t* wire,
+                                                    uint32_t count,
+                                                    uint32_t cipher_suite_len,
+                                                    struct s2n_cipher_suite** negotiated_cipher) {
+    RESULT_ENSURE_REF(wire);
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(negotiated_cipher);
+    RESULT_ENSURE_EQ(*negotiated_cipher, NULL);
+
+    const struct s2n_security_policy *security_policy;
+    RESULT_GUARD_POSIX(s2n_connection_get_security_policy(conn, &security_policy));
+    const struct s2n_cipher_preferences* cipher_preferences = security_policy->cipher_preferences;
+    RESULT_ENSURE_REF(cipher_preferences);
+
+    bool in_group = false;
+    /** 
+     * Tracks mutually supported cipher suite in the client index. This is needed to track tie-breakers
+     * in an equal-preference group.
+     */
+    int negotiated_client_index = count;
+    /** 
+     * The highest protocol version match index (also a server index). This is a fall back index if
+     * no other mutually-supported cipher suite is identified.
+     */
+    int highest_vers_match_index = -1;
```

Comment:
I would change highest_vers_match_index to be a cipher suite pointer instead of an index. (Don't forget the comment too)

Reply:
Will change